test/Test/Database/LSMTree/Internal/Snapshot/Codec/Golden.hs

```diff
@@ -28,329 +30,490 @@ import qualified System.FS.API as FS
+    ]
 
-enumerateSnapMergingRun ::
-     [(ComponentAnnotation, t)]
-  -> [(ComponentAnnotation, SnapMergingRun t SnapshotRun)]
-enumerateSnapMergingRun mTypes =
-    [ (fuseAnnotations ["C0", blank, blank],
-       SnapCompletedMerge mergeDebt enumerateOpenRunInfo)
-    | mergeDebt <- (MR.MergeDebt. MR.MergeCredits) <$> [ magicNumber2 ]
+{-------------------------------------------------------------------------------
+  Enumeration class
+-------------------------------------------------------------------------------}
+
+-- | Enumerate values of type @a@ for golden tests
+--
+-- To prevent combinatorial explosion, the enumeration should generally be
+-- /shallow/: the different constructors for type @a@ should be enumerated
+-- without recursively enumerating the constructors' fields. For example,
+-- enumerating @Maybe Int@ should give us something like:
+--
+-- > enumGolden @(Maybe Int) = [ Just 17, Nothing ]
+--
+-- This is generally a suitable approach, since the snapshot encodings do not
+-- encode types differently depending on values in the constructor fields.
+--
+-- Example (recursive) instances that prevent combinatorial explosion:
+--
+-- @
+--  instance EnumGolden a => EnumGolden (Maybe a) where
+--    enumGolden = [ Just (singGolden @a), Nothing ]
+--  instance EnumGolden Int where
+--    enumGolden = [17, -72] -- singGolden = 17
+-- @
+--
+-- If there are encoders that do require more elaborate (recursive)
+-- enumerations, then it is okay to deviate from shallow enumerations, but take
+-- care not to explode the combinatorics ;)
+class EnumGolden a where
+  {-# MINIMAL enumGolden | enumGoldenAnnotated | singGolden #-}
+
+  -- | Enumerated values. The enumeration should be /shallow/.
+  --
+  -- The default implementation is to return a singleton list containing
+  -- 'singGolden'.
+  enumGolden :: [a]
+  enumGolden = [ singGolden ]
+
+  -- | Enumerated values with an annotation for naming purposes. The enumeration
+  -- should be /shallow/, and the annotations should be unique.
+  --
+  -- The default implementation is to annotate 'enumGolden' with capital letters
+  -- starting with @\'A\'@.
+  enumGoldenAnnotated :: [(Annotation, a)]
+  enumGoldenAnnotated = zip [[c] | c <- ['A' .. 'Z']] enumGolden
+
+  -- | A singleton enumerated value. This is mainly useful for superclass
+  -- instances.
+  --
+  -- The default implementation is to take the 'head' of 'enumGoldenAnnotated'.
+  singGolden :: a
+  singGolden = snd $ head enumGoldenAnnotated
+
+type Annotation = String
+
+enumGoldenAnnotated' :: EnumGolden a => Proxy a -> [(Annotation, a)]
+enumGoldenAnnotated' _ = enumGoldenAnnotated
+
+{-------------------------------------------------------------------------------
+  Enumeration class: names and file paths
+-------------------------------------------------------------------------------}
+
+nameGolden :: Typeable a => Proxy a -> Annotation -> String
+nameGolden p ann = map spaceToUnderscore (show $ typeRep p) ++ "." ++ ann
```

Comment:
The downside of using `typeRep` is that the name of the golden file/test changes when the type name changes. In some cases, we might want to rename the type without affecting the name, so maybe `nameGolden` should be a member of `EnumGolden`?

Reply:
The `typeRep` approach can still be the default, but the upside is that you can change the name per instance

Reply:
Would this be an issue? Seems easier to just recreate the relevant golden files than to maintain a type-name mapping forever.

Reply:
Maybe not, it's just something that came to mind. I'm fine with leaving it as is too